compiler/rustc_hir_analysis/src/check/always_applicable.rs

```diff
@@ -0,0 +1,195 @@
+use rustc_data_structures::fx::FxHashSet;
+use rustc_errors::{struct_span_err, ErrorGuaranteed};
+use rustc_infer::infer::outlives::env::OutlivesEnvironment;
+use rustc_infer::infer::{RegionResolutionError, TyCtxtInferExt};
+use rustc_middle::ty::subst::SubstsRef;
+use rustc_middle::ty::util::CheckRegions;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_trait_selection::traits::{self, ObligationCtxt};
+
+use crate::hir::def_id::DefId;
+
+/// This function confirms that the trait implementation identified by
+/// `trait_impl_def_id` is not any more specialized than the type it is
+/// attached to (Issue #8142).
+///
+/// This means:
+///
+/// 1. The generic region/type parameters of the impl's self type must
+///    all be parameters of the Trait impl itself (i.e., no
+///    specialization like `impl Trait for Foo<i32>`), and,
+///
+/// 2. Any bounds on the generic parameters must be reflected in the
+///    struct/enum definition for the nominal type itself (i.e.
+///    cannot do `struct S<T>; impl<T:Clone> Trait for S<T> { ... }`).
+///
+pub fn check_trait_impl(tcx: TyCtxt<'_>, trait_impl_def_id: DefId) -> Result<(), ErrorGuaranteed> {
+    let self_type = tcx.type_of(trait_impl_def_id).subst_identity();
+    check_trait_impl_given_ty(tcx, trait_impl_def_id, self_type)
+}
+
+fn check_trait_impl_given_ty<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    trait_impl_def_id: DefId,
+    ty: Ty<'tcx>,
+) -> Result<(), ErrorGuaranteed> {
+    match ty.kind() {
+        ty::Adt(adt_def, adt_to_impl_substs) => {
+            ensure_trait_params_and_item_params_correspond(
+                tcx,
+                trait_impl_def_id,
+                adt_def.did(),
+                adt_to_impl_substs,
+            )?;
+
+            ensure_trait_predicates_are_implied_by_item_defn(
+                tcx,
+                trait_impl_def_id,
+                adt_def.did(),
+                adt_to_impl_substs,
+            )
+        }
+
+        ty::Ref(_, ty, _) | ty::Array(ty, _) | ty::Slice(ty) => {
+            check_trait_impl_given_ty(tcx, trait_impl_def_id, *ty)
+        }
+
+        ty::Tuple(tys) => {
```

Comment:
Similarly, I would've thought that tuples kinda act like an ADT like:

```
struct TupleN<A, B, ..., Z: ?Sized>(A, B, ..., Z);
//~^ Last field may be unsized...
```

so you could only implement always-applicable negative impls like `impl<A, B, C: ?Sized> !Trait for (A, B, C)`.